homie-controller/src/lib.rs

```diff
@@ -248,6 +252,82 @@ impl HomieController {
                 device.implementation = Some(payload.to_owned());
                 Some(Event::device_updated(device))
             }
+            [device_id, "$extensions"] => {
+                let device = get_mut_device_for(devices, "Got extensions for", device_id)?;
+                device.extensions = payload
+                    .split(",")
+                    .map(|part| part.parse())
+                    .collect::<Result<Vec<_>, _>>()?;
+                Some(Event::device_updated(device))
+            }
+            [device_id, "$localip"] => {
+                let device = get_mut_device_for(devices, "Got localip for", device_id)?;
+                device.local_ip = Some(payload.to_owned());
+                Some(Event::device_updated(device))
+            }
+            [device_id, "$mac"] => {
+                let device = get_mut_device_for(devices, "Got mac for", device_id)?;
+                device.mac = Some(payload.to_owned());
+                Some(Event::device_updated(device))
+            }
+            [device_id, "$fw", "name"] => {
+                let device = get_mut_device_for(devices, "Got fw/name for", device_id)?;
```

Comment:
thinking about this further, I'd be slightly more happy with something of the form `let device = get_mut_device_for_update(devices, device_id, "$fw/name")?;`

If we go down the macro route then we could even do the error message formatting in the macro and kill off get_mut_device_for() entirely.

Reply:
I was thinking of doing that originally, but then it doesn't work quite so well calling it from `get_mut_node_for`.